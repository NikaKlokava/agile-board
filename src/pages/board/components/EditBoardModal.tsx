import { Button } from "../../../shared/components/button";
import { FieldWrapper } from "../../../shared/components/field_wrapper";
import { Input } from "../../../shared/components/input";
import { ModalWrapper } from "../../../shared/components/modal_wrapper";
import cl from "./styles/edit_board_modal.module.css";

export const EditBoardModal = () => {
  return (
    <ModalWrapper>
      <h2 className={cl.add_task_title}>Edit Board</h2>
      <FieldWrapper fieldName="Board Name">
        <Input withDelete={false} />
      </FieldWrapper>
      <FieldWrapper fieldName="Board Columns" clName="style_container">
        <Input withDelete={true} />
        <Input withDelete={true} />
        <Input withDelete={true} />
      </FieldWrapper>
      <Button text={"Add New Column"} withIcon={true} />
      <Button text={"Save Changes"} withIcon={false} newClass="center" />
    </ModalWrapper>
  );
};
