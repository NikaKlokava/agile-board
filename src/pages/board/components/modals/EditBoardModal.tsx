import { FieldArray, Form, Formik } from "formik";
import { useDispatch, useSelector } from "react-redux";
import { addNewColumn } from "../../../../redux/actionCreators/newBoardCreator";
import { Button } from "../../../../shared/components/button";
import { FieldName } from "../../../../shared/components/field_name";
import { Input } from "../../../../shared/components/input";
import { ModalWrapper } from "../../../../shared/components/modal_wrapper";
import { EditBoardSchema } from "../../../../utils/utils";
import { AddBtn } from "../../../../shared/components/add_button";
import cl from "./modal_styles.module.css";

type Props = {
  onClose: () => void;
};

export const EditBoardModal = ({ onClose }: Props) => {
  const activeBoard = useSelector<RootState, BoardType>(
    (state) => state.activeBoard
  );

  const dispatch = useDispatch();

  const onSubmit = (values: BoardType) => {
    const columns = values.columns.filter(
      (column) => column?.title && column.title.trimStart().length !== 0
    );
    dispatch(addNewColumn(values.uuid, values.name, columns));
    onClose();
  };

  const initialData = {
    name: activeBoard.name,
    columns: activeBoard.columns.reduce((accum: string[], current) => {
      return [...accum, current.title];
    }, []),
    uuid: activeBoard.uuid,
  };

  const uuids = activeBoard.columns.reduce((accum: string[], current) => {
    return [...accum, current.uuid];
  }, []);

  return (
    <ModalWrapper onWrapperClick={onClose}>
      <Formik
        initialValues={initialData}
        validationSchema={EditBoardSchema}
        onSubmit={(values) => {
          const columns = values.columns.filter(
            (column) => column && column.trimStart().length !== 0
          );
          const newColumns = columns.reduce(
            (accum: { title: string; uuid: string }[], current, index) => {
              return [...accum, { title: current, uuid: uuids[index] }];
            },
            []
          );
          onSubmit({
            ...values,
            columns: newColumns,
          });
        }}
      >
        {(props) => (
          <Form
            onSubmit={(e) => {
              e.preventDefault();
              props.handleSubmit();
            }}
            className={cl.form_container}
          >
            <h2 className={cl.modal_title} data-testid="edit-board-modal">
              Edit Board
            </h2>
            <FieldName formikName={"name"} />
            {props.errors.name && props.touched.name && (
              <p style={{ color: "red" }}>{props.errors.name}</p>
            )}
            <FieldArray
              name="columns"
              render={(arrayHelpers) => (
                <>
                  <div className={cl.container}>
                    <p className={cl.title}>Board Columns</p>
                    {props.values.columns.map((_, index) => (
                      <Input
                        key={index}
                        formikName={`columns[${index}]`}
                        remove={arrayHelpers.remove}
                        index={index}
                        uuids={uuids}
                      />
                    ))}
                  </div>
                  <AddBtn
                    add={arrayHelpers.push}
                    text="Add New Column"
                    testid={"add-new-column-btn"}
                  />
                </>
              )}
            />
            <Button
              text={"Save Changes"}
              withIcon={false}
              newClass="center"
              testid={"save-changes-btn"}
              type="submit"
            />
          </Form>
        )}
      </Formik>
    </ModalWrapper>
  );
};
