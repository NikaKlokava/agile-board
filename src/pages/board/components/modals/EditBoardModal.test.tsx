import renderer from "react-test-renderer";
import { EditBoardModal } from "./EditBoardModal";

describe("Test the EditBoardModal component", () => {
  test("The EditBoardModal renders correctly", () => {
    const editBoardModalSnap = renderer.create(<EditBoardModal />).toJSON();
    expect(editBoardModalSnap).toMatchSnapshot();
  });
});
