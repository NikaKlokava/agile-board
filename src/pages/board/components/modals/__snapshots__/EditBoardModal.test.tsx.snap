// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the EditBoardModal component The EditBoardModal renders correctly 1`] = `
<div
  className="modal_wrapper"
>
  <div
    className="modal_container"
  >
     
    <h2
      className="modal_title"
    >
      Edit Board
    </h2>
    <div
      className="container"
    >
      <p
        className="title"
      >
        Board Name
      </p>
      <div
        className="input_container"
      >
        <input
          className="input_style"
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
      </div>
    </div>
    <div
      className="container style_container"
    >
      <p
        className="title"
      >
        Board Columns
      </p>
      <div
        className="input_container"
      >
        <input
          className="input_style"
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
        <div
          className="delete_icon"
        />
      </div>
      <div
        className="input_container"
      >
        <input
          className="input_style"
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
        <div
          className="delete_icon"
        />
      </div>
      <div
        className="input_container"
      >
        <input
          className="input_style"
          placeholder="e.g Take coffee break"
          spellCheck={false}
          type="text"
        />
        <div
          className="delete_icon"
        />
      </div>
    </div>
    <div
      className="button"
    >
      <div
        className="button_icon"
      />
      <div
        className="button_text"
      >
        Add New Column
      </div>
    </div>
    <div
      className="button center"
    >
      <div
        className="button_text"
      >
        Save Changes
      </div>
    </div>
  </div>
</div>
`;
